markdown: Allow bold/italic formatting within links.

adisingh396 opened the pull request:
This pull request fixes a long-standing bug where inline Markdown formatting (like bold and italic) did not render inside of links.

The root cause was that the link text was intentionally wrapped in an `AtomicString` to prevent other custom rules (like issue linkifiers) from creating invalid nested `<a>` tags.

The fix involves two main changes:
* The `AtomicString` wrapper is removed from the `LinkInlineProcessor`, which allows standard inline formatting like **bold** and *italic* to be processed within link text.
* A new `UnnestLinks` treeprocessor is introduced. It runs after all other processors and safely cleans up any accidentally nested <a> tags that might now be created by linkifiers. This ensures the final HTML is valid while allowing the desired formatting to work.

A new test case was added to verify the fix and prevent regressions.

Fixes: #36087 

**Screenshots and screen captures:**
#### Before
![Before](https://github.com/user-attachments/assets/b0f8a88b-0c9d-4f4d-928d-517a6f4e0194)
#### After
![After](https://github.com/user-attachments/assets/3cb8ce45-08be-4494-8cfc-24939799da90)

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2024,9 +2070,11 @@ def zulip_specific_link_changes(self, el: Element) -> None | Element:
+        # which can result in nested links. We've removed this in favor of the
+        # `UnnestLinks` treeprocessor, which allows for other inline markdown
+        # to render (e.g. bold and italics) and cleans up any nested links later.
+        # el.text = markdown.util.AtomicString(el.text)
```
We'll want to see tests that cover the type of corner cases that this previous code was meant to cover.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -575,6 +575,52 @@ def run(self, root: Element) -> None:
+                        # Finally, remove the nested_a tag from the tree.
+                        parent.remove(nested_a)
+                        # We found and processed it, so we can break the inner loop.
+                        break
```
This is a pretty complicated bit of logic. We'll want to see a proper test suite for it handling the types of corner cases that it's designed to address.

I'm overall a bit skeptical of the approach of generating invalid HTML and then trying to fix it up. Usually, it's best to avoid generating invalid states in the first place. One idea might be to have a subclass or variant of `AtomicString` that does allow certain processing, but I'm skeptical that can be done simply. Best would be to figure out if there's a way to manage the ordering of steps or how linkifiers work to protect against the failure mode in question directly.

zendrix396 commented:
Hey @timabbott, thanks for the thorough review and feedback! Appreciate it.

I agree that we need tests for the corner cases that `AtomicString` was protecting against. I added a few tests (link without text, link with empty text, and link containing a linkifier and autolink pattern) that would fail with nested links if the `UnnestLinks` processor wasn’t present. And they all pass.

I understand your skepticism about the “generate and fix” approach. I know that preventing the invalid HTML in the first place would be the cleaner solution. But it is kind of tricky because the inline patterns run in a fixed sequence and there isn’t an easy way to tell the linkifier/autolink patterns “you are currently inside the text of another link, so don’t run”. The `UnnestLinks` treeprocessor seemed like the most direct way to solve the problem within the existing structure.

I’m open to other solutions if you have one. For now I pushed the new tests to make this more robust.

Thanks again!

adisingh396 committed:
markdown: Allow bold/italic formatting within links.

Previously, bold and italic markdown was not rendered inside
links because the link text was wrapped in 'AtomicString' to
prevent nested linkification.

This commit introduces a 'SemiAtomicString' class. Standard
markdown emphasis processes this string normally, but custom
Zulip linkifiers (like stream links and mentions) detect it
and abort, cleanly preventing nested <a> tags.

To ensure this executes correctly, Zulip's custom linkifiers
were shifted to a priority below LinkInlineProcessor (60),
guaranteeing the Markdown link wrapper runs first.

Fixes: #36087

adisingh396 commented:
hey @timabbott, following up on your feedback I tried replacing the previous generate and fix approach with a cleaner alternative of creating a separate `SemiAtomicString` class as a variant of original `AtomicString`. Standard emphasis formats it normally, but our custom `ZulipInlineProcessor` detects it and explicitly aborts any pattern flagged with `is_link_pattern=True`.
For this to work, I shifted Zulip's custom link-generating patterns (stream mentions, user mentions, etc.) to run at a priority just subtly below the standard `LinkInlineProcessor` (60).
Now standard markdown links parse _first_ and wrap their inner text in `SemiAtomicString`. This will kind of physically prevent invalid nested HTML from ever being generated. I've also removed previous `UnnestLinks` script.
